### DISCUSSION TITLE
Device tracker vs SNMP?

### DISCUSSION THREAD

**Discussion by nagyrobi** [travisghansen/hass-pfsense#6] [General]
In what aspect is this implementation of device tracker better than [HA's built-in SNMP](https://www.home-assistant.io/integrations/snmp/#precense-detection) working with pfSense?

**Comment by travisghansen** [comment]
It probably isn’t. I’ve never used snmp but it looks like they both would do the exact same thing if they are looking at the arp table.

**Reply by nagyrobi** [reply]
One problem of the SNMP integration is that it's too generic, it adds an entity for each item found in the ARP table which is an overkill. Having ~100 MAC address entities in HA just for monitoring 2 of them is spamming the entity database...

In this component, would it be possible to turn things around? Just specify at component configuration a list of MAC addresses to monitor (optionally with a name) and only add those as binary sensor entities. And set to true if present in ARP, and false if not.

**Reply by travisghansen** [reply]
Anything is possible. I think what I should probably do is just disable by default the entity and let the user enable the entries they care about.

One thing I have considered is allowing a configuration option that forces a refresh of the arp entry after some specific interval. I suppose the user could alter the default on the os but that may be undesirable for certain clients.

**Reply by nagyrobi** [reply]
The main problem I have that there are "Guest" networks where hundreds, even thousands of visiting clients connect just on events/occasions for a limited period of time and they never come back. This makes the number of unneeded MAC addresses continuously grow, not to speak about MAC randomization which is used by many mobile devices these days (same device uses a different random address each time it connects again).
Are the entities corresponding to such MAC addresses being removed after some time?

**Reply by travisghansen** [reply]
Currently no entities are removed. That could prove difficult as it may be hard to determine when something is simply no longer present vs should be removed permanently.

Regarding randomization, to my knowledge the randomness happen at the ap level not at the connection level. ie: connecting to the same ap should always use the same mac but across aps the mac will change.

All great thoughts! I want to keep the config as simplified as possible but these are good points to bring up.

**Reply by nagyrobi** [reply]
> Regarding randomization, to my knowledge the randomness happen at the ap level not at the connection level. ie: connecting to the same ap should always use the same mac but across aps the mac will change.

True, but when you have a venue with 50 APs all in the same VLAN...

**Reply by travisghansen** [reply]
Sure fair enough. I was under the impression it was based on the ssid and not bssid but I could be off entirely!

**Reply by travisghansen** [reply]
Do you happen to know if it's possible to sanely handle a list of entries for a single voluptuous option? I'm wondering if maybe we should add a config option that lets the user input a list of mac regexes or something of that nature.

**Reply by nagyrobi** [reply]
https://github.com/nagyrobi/home-assistant-custom-components-linkplay/blob/65f3beebd2caa36b77f43db6021c5b79e81522f3/custom_components/linkplay/media_player.py#L128

I specify a list then:
https://github.com/nagyrobi/home-assistant-custom-components-linkplay#configuration

I don't do the flow stuff, but under the hood should be the same.

**Reply by travisghansen** [reply]
I don't think that will work as the list of available options if finite and handled as a dict. In this case we just want to let the user input a string (mac address) and append it to the list, and then later potentially remove the item from the list.

**Reply by nagyrobi** [reply]
Nested dict maybe?

**Reply by travisghansen** [reply]
Good question…I’ll mess with it a little bit more and see where things land.

**Reply by travisghansen** [reply]
FYI the integration now interacts with the arp table and actually deletes the arp entry every `poll interval` which means the data is consistent and never stale. That likely provides an experience superior to the snmp data.

**Reply by nagyrobi** [reply]
Doesn't that interfere with other functionality on the firewall?

**Comment by travisghansen** [comment]
Not really. The arp entry by default is removed every 20 minutes anyhow so it’s only happening more frequently (and notably only for the selected devices, not the whole arp table).

**Comment by nagyrobi** [comment]
Works almost perfect. Would be nice an extra parameter, a configurable time period within the device tracked is being considered at home (bigger or equal with device poll time), regardless of what's in the arp table. My wife's phone keeps flapping between home and away, I guess because of it powersaving on the wifi module. Can't be configured to act differently.

The logic would be, if the arp poll says it's offline, and it doesn't come back within the time period configured above (time expires and still away), only _then_ set the entity to away.

**Reply by nagyrobi** [reply]
Also would be nice to enable history for the tracker entities, could be helpful to visualise flapping.

**Reply by travisghansen** [reply]
Why not just increase the poll time? It's not exactly the same as what you describe, but pretty close.

History should be managed automatically by HA.

**Reply by nagyrobi** [reply]
Problem that with higher poll times I will get even longer periods of times when the device is falsely offline. So small poll times are needed +a timeout to hop over the flaps.

**Reply by travisghansen** [reply]
Yeah that's fair. With the phone what's the amount of time between 'flaps' when it's asleep? When it's asleep does it ever come back without being woke up?

**Reply by nagyrobi** [reply]
Lemme debug why I don't get history for the tracker entities, I think I may have disabled them a while ago because of some integtations spamming the database...

**Reply by travisghansen** [reply]
Yeah, and if you added the integration a couple weeks ago everything in the arp table was getting added. Now only explicitly included devices are being added as entities (I would recommend removing the firewall and re-adding to start fresh).

**Comment by nagyrobi** [comment]
![Screenshot_20211114-205928](https://user-images.githubusercontent.com/1550668/141696602-6fad4c23-c37a-4568-93d4-fa51f0de3a45.jpg)

**Reply by nagyrobi** [reply]
I would use a 3 or 5 minute timeout period to consider the device at home, even if it's reported as away. Polling time of 1 minute is enough.

**Reply by travisghansen** [reply]
Out of curiosity what kind of device is that?

**Reply by nagyrobi** [reply]
It's a phone running Android 10.
It's not the only one, other family member's phones behave similar.

**Reply by nagyrobi** [reply]
The old SNMP integration also has `consider_home` parameter to mitigate this:
https://www.home-assistant.io/integrations/device_tracker/

**Reply by travisghansen** [reply]
Any iOS devices?

Understood about the parameter for the other integrations.

**Reply by nagyrobi** [reply]
No iOS here...

**Reply by nagyrobi** [reply]
Did you get a chance to consider implementing a `consider_home` parameter?

**Reply by travisghansen** [reply]
Actually was planning to look at it today :)

**Reply by travisghansen** [reply]
Give it a whirl https://github.com/travisghansen/hass-pfsense/commit/0662246e2b525c31807eda0ebbb00a02ffba83a8

**Reply by nagyrobi** [reply]
Installed, trying with value set to 600

**Reply by nagyrobi** [reply]
A whole lot better since this is possible:
![kép](https://user-images.githubusercontent.com/1550668/146983569-ef7b17e4-c6a5-4bba-b98b-79d9cc079c0e.png)

Is it in the latest release?

**Reply by travisghansen** [reply]
It's in `main` but I haven't tagged a versioned release with it yet no.

**Reply by travisghansen** [reply]
I've added some openvpn sensors in #43 and am just waiting for some feedback on that. After that looks clean/clear I'll probably create a new versioned release.

**Reply by travisghansen** [reply]
Still holding strong?

**Reply by nagyrobi** [reply]
Absolutely. This was needed here.
